javascriptlib/src/mill/javascriptlib/TypeScriptModule.scala

```diff
@@ -1,83 +1,125 @@
+    Task { Map("*" -> sources().path.toString(), "*" -> npmInstall().path.toString()) }
 
   def compile: T[(PathRef, PathRef)] = Task {
-    val nodeTypes = npmInstall().path / "node_modules/@types"
     val javascriptOut = Task.dest / "javascript"
     val declarationsOut = Task.dest / "declarations"
 
     val upstreamPaths =
-      for (((jsDir, dTsDir), mod) <- Task.traverse(moduleDeps)(_.compile)().zip(moduleDeps))
+      for (((_, dTsDir), mod) <- Task.traverse(moduleDeps)(_.compile)().zip(moduleDeps))
         yield (mod.millSourcePath.subRelativeTo(Task.workspace).toString + "/*", dTsDir.path)
 
-    val allPaths = upstreamPaths ++ Seq("*" -> sources().path, "*" -> npmInstall().path)
+    val combinedPaths = upstreamPaths ++ tsCompilerOptionsPaths().toSeq
+    val combinedCompilerOptions: Map[String, ujson.Value] = Map(
+      "declarationDir" -> ujson.Str(declarationsOut.toString),
+      "outDir" -> ujson.Str(javascriptOut.toString),
+      "paths" -> ujson.Obj.from(combinedPaths.map { case (k, v) => (k, ujson.Arr(s"$v/*")) })
+    ) ++ tsCompilerOptions()
 
     os.write(
       Task.dest / "tsconfig.json",
       ujson.Obj(
-        "compilerOptions" -> ujson.Obj(
-          "outDir" -> javascriptOut.toString,
-          "declaration" -> true,
-          "declarationDir" -> declarationsOut.toString,
-          "typeRoots" -> ujson.Arr(nodeTypes.toString),
-          "paths" -> ujson.Obj.from(allPaths.map { case (k, v) => (k, ujson.Arr(s"$v/*")) })
-        ),
+        "compilerOptions" -> ujson.Obj.from(combinedCompilerOptions.toSeq),
         "files" -> allSources().map(_.path.toString)
       )
     )
 
-    os.call((npmInstall().path / "node_modules/typescript/bin/tsc"))
+    os.call(npmInstall().path / "node_modules/typescript/bin/tsc")
 
     (PathRef(javascriptOut), PathRef(declarationsOut))
   }
 
-  def mainFileName = Task { s"${millSourcePath.last}.js" }
+  def mainFileName: Target[String] = Task { s"${millSourcePath.last}.js" }
+
+  def mainFilePath: Target[Path] = Task { compile()._1.path / mainFileName() }
+
+  def mkENV =
+    Task.Anon { Map("NODE_PATH" -> Seq(".", compile()._1.path, npmInstall().path).mkString(":")) }
 
-  def prepareRun = Task.Anon {
+  def prepareRun: Task[(Path, Map[String, String])] = Task.Anon {
     val upstream = Task.traverse(moduleDeps)(_.compile)().zip(moduleDeps)
-    for (((jsDir, tTsDir), mod) <- upstream) {
+    for (((jsDir, _), mod) <- upstream) {
       os.copy(jsDir.path, Task.dest / mod.millSourcePath.subRelativeTo(Task.workspace))
     }
-    val mainFile = compile()._1.path / mainFileName()
-    val env = Map("NODE_PATH" -> Seq(".", compile()._1.path, npmInstall().path).mkString(":"))
-    (mainFile, env)
+    (mainFilePath(), mkENV())
   }
 
-  def run(args: mill.define.Args) = Task.Command {
+  // define computed arguments and where they should be placed (before/after user arguments)
+  def computedArgs: Task[(Option[Seq[String]], Option[Seq[String]])] = Task { (None, None) }
```

Comment:
We don't seem to use the "after" case at all, let's just drop it and make this return a `Task[Seq[String]]` as I said before

Reply:
okay will make the changes in basic-1